### ISSUE TITLE
BUG: RegularGridInterpolator is incredibly slow compared to interp2d

### ISSUE THREAD

**Issue by sjev** [scipy/scipy#18010]
### Describe your issue.

`interp2d` is being depricated. `RegularGridInterpolator` is being suggested as replacement.
However, the new code is 10+ times slower with linear interpolation and 1000+ with cubic.

reported on [stackoverflow](https://stackoverflow.com/questions/75427538/regulargridinterpolator-excruciatingly-slower-than-interp2d)

### Reproducing Code Example

```python
#!/usr/bin/env python3
"""
compare scipy.interpolate.interp2d and scipy.interpolate.RegularGridInterpolator
"""

import numpy as np
import time
from scipy.interpolate import RegularGridInterpolator
from scipy.interpolate import interp2d

import warnings

warnings.filterwarnings("ignore", category=DeprecationWarning)


# --------- interp2d
x = np.linspace(0, 4, 13)
y = np.array([0, 2, 3, 3.5, 3.75, 3.875, 3.9375, 4])
X, Y = np.meshgrid(x, y)
Z = np.sin(np.pi * X / 2) * np.exp(Y / 2)

x2 = np.linspace(0, 4, 65)
y2 = np.linspace(0, 4, 65)
f = interp2d(x, y, Z, kind="cubic")

t_start = time.perf_counter()
Z2 = f(x2, y2)
t_end = time.perf_counter()
print(f"interp2d: {t_end - t_start:.3e} s")


# --------- RegularGridInterpolator
X, Y = np.meshgrid(x, y, indexing="ij")  # indexing="ij" is important
Z = np.sin(np.pi * X / 2) * np.exp(Y / 2)
f2 = RegularGridInterpolator((x, y), Z, method="cubic")

X2, Y2 = np.meshgrid(x2, y2)
t_start = time.perf_counter()
Z3 = f2((X2, Y2))
t_end = time.perf_counter()
print(f"RegularGridInterpolator: {t_end - t_start:.3e} s")

# check that the two methods give the same result
assert np.allclose(Z2, Z3)


```

### Error message

```shell
interp2d: 1.102e-03 s
RegularGridInterpolator: 1.337e+00 s

```

### SciPy/NumPy/Python version information

1.10.0 1.24.2 sys.version_info(major=3, minor=10, micro=8, releaselevel='final', serial=0)

**Comment by ev-br** [comment]
Yes, it's sadly a known issue. For the linear mode the difference for me is not as drastic. Using a slight MVEing of the OP below, I get locally a difference of about 2.5x. Which is still a lot, of course. Seeing what we can do about it is in the cards, https://github.com/scipy/scipy/issues/17356 tracks it.

Cubic modes are known to be much slower, and we're working on it. Some preliminary work: https://github.com/ev-br/scipy/tree/ndbspline

Help moving either of these two forward is appreciated.

```
In [4]: import numpy as np
   ...: import matplotlib.pyplot as plt
   ...: from scipy.interpolate import RegularGridInterpolator
   ...: from scipy.interpolate import interp2d
   ...:
   ...: import warnings
   ...: warnings.filterwarnings("ignore", category=DeprecationWarning)
   ...:
   ...:
   ...:
   ...: # %% compare performance
   ...:
   ...:
   ...: x = np.linspace(0, 4, 13)
   ...: y = np.array([0, 2, 3, 3.5, 3.75, 3.875, 3.9375, 4])
   ...: X, Y = np.meshgrid(x, y)
   ...: Z = np.sin(np.pi*X/2) * np.exp(Y/2)
   ...:
   ...: x2 = np.linspace(0, 4, 65)
   ...: y2 = np.linspace(0, 4, 65)
   ...: f = interp2d(x, y, Z, kind='linear')
   ...:
   ...: %timeit f(x2, y2)
   ...:
   ...: X, Y = np.meshgrid(x, y, indexing='ij', sparse=True)
   ...: Z = np.sin(np.pi*X/2) * np.exp(Y/2)
   ...: f2 = RegularGridInterpolator((x, y), Z, method='linear')
   ...:
   ...: %timeit f2((x2, y2))
32 µs ± 219 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
79.7 µs ± 557 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

**Comment by sjev** [comment]
I would suggest **not deprecating** `interp2d`, it's widely used and does a good job at what it is supposed to do. Please don't break legacy code with a new interface and decreased performance.

**Comment by ev-br** [comment]
The discussion of this was going from October 2022: https://mail.python.org/archives/list/scipy-dev@python.org/thread/LU72CAGYFH37PZWCANMUF6UWFNSPZRHU/#O5PF6MPP2JKZVDBMKTR6E4QM3NDPEWKA , that email thread lists reasons for the deprecation.

At this point, if you want bit-to-bit compatibility and performance, you can reuse the fitpack routines which interp2d wraps.

**Comment by tupui** [comment]
I think we can close this. There is nothing here we are not tracking elsewhere.

As for the legacy/deprecation part. The ship has sailed to me as we did consult the community and discussed pro/cons.

**Comment by sjev** [comment]
Hi guys, I've read the discussion from October 2022. While I do understand the decision to deprecate `interp2d` ,  I'm afraid that a lot of users will be hurt or surprised in an unplesant way. I really appreciate your efforts to improve scipi, but in this case it feels that the cure is worse than the desease.
My personal experience - I saw the deprecation notice and decided to update my code. There was no clear guide anywhere on how to do it.  It took me a couple of hours to understand the new interface and port just a couple of scipts. mainly because of the things like custom indexing `X, Y = np.meshgrid(x, y, indexing="ij", creating  meshgrids where I previously used  vectors etc.)`.  Then the performance issue popped up. 
In many cases the performance loss would be acceptable, but I often use this function in video stream processing and it's hard for me to accept any performance loss there.
To be honest, I just need simple linear interpolation to correct 2d grids. Something 'matlab-like' ;-).  
For me personally, it would be fine if `interp2d` would be stripped internally to linear interpolation, as long at there is no significant performance loss. Or if there is a clear porting guide or a wrapper, again as long as there is no regression on performance.
Regarding using fitpack routines - well I don't mind spending an hour or two writing my own new wrapper, but at this moment I don't even know what fitpack is, so that hole will probably go much deeper than I'd like or could manage...
So please keep in mind that the users of scipy did not read the discussion last October, so 'we did consult the community'   is a bit of an overstatement imo. 
**To sum up:** as long as there is no clear porting guide to a solution without performance regression, `interp2d` should not be deprecated.   Doing this will frustrate a lot of users and give companies like Matlab arguments to advocate against open source, lie "things can change and break all of a sudden"

**Comment by tupui** [comment]
> but in this case it feels that the cure is worse than the desease

Of course, this is your point of view as a user. Which is important, but as maintainer of said code, we have multiple perspectives to balance.

> So please keep in mind that the users of scipy did not read the discussion last October, so 'we did consult the community' is a bit of an overstatement imo.

We are aware that our system is not perfect. We are making more outreach activity to interact with users.

Still, people involved in the discussions on the mailing list and GitHub are quite experienced. We do our best to make informed and sensible decisions which have the community's best interest in mind.

> To sum up: as long as there is no clear porting guide to a solution without performance regression, interp2d should not be deprecated. Doing this will frustrate a lot of users and give companies like Matlab arguments to advocate against open source, lie "things can change and break all of a sudden"

This is not how we proceed. At some point we need to make some judgment calls to move forward. Otherwise we do nothing. At SciPy's scale, doing literally anything is breaking people.

For the current regression, as Evgeni said we are tracking it and hope to have a solution by the time the deprecation period expire. If more reports gets in and the community really comes forward and we flag this as vital, we will of course take it into account. For now, this is not the case.

We strive to provide good documentation. The situation really improved in the last years. It's still hard to keep up and we do our best here. As we said here, all that is known and we "just" need to do the work. Any help is much appreciated as always.

**Comment by ev-br** [comment]
Thanks for sharing your perspective. Much appreciated! 

Re: porting guide: I get the feeling that you haven't seen https://gist.github.com/ev-br/8544371b40f414b7eaf3fe6217209bff which lists precise fitpack equivalents (FITPACK is the Fortran library from the 80s which interp2d delegates the work to). So the questions for the porting guide are
- is this gist clear? If not, how to improve it? Ideally, send a pull request; if that's hard, can put the desired changes here.
- how this gist could be made more discoverable? It currently is in the deprecation message, the link is straight from the red box in https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interp2d.html
Suggestions to where to put the link so that users see it are greatly appreciated

Also note that for the time being you can just silence the DeprecationWarning. Given that interp2d is indeed quite widely used, the plan is to have an extended deprecation period. But that I personally would not recommend.

**Comment by sjev** [comment]
Thank you for the pointers!
I did not see the gist, as the link in deprecaton notice seemed to be broken, so I just started googling. It does work now btw. 
As long as the link keeps working, I think it's fine this way.  
I'll give it another go in coming days.  
One suggestion for the gist - as introduction a short description on why the `interp2d` is being deprecated would be very welcome.

**Comment by j-bowhay** [comment]
We have been slightly bitten by https://github.com/pydata/pydata-sphinx-theme/pull/877 I believe as we are not using the most recent theme version (@tupui please correct me if this wrong). This means you cannot click the link to the gist in the documentation, see https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interp2d.html#scipy.interpolate.interp2d

**Comment by ev-br** [comment]
Tried fixing the link in https://github.com/scipy/scipy/pull/18022, not sure if it worked or what does it take it to make it work.

**Comment by tupui** [comment]
> We have been slightly bitten by [pydata/pydata-sphinx-theme#877](https://github.com/pydata/pydata-sphinx-theme/pull/877) I believe as we are not using the most recent theme version (@tupui please correct me if this wrong). This means you cannot click the link to the gist in the documentation, see [docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interp2d.html#scipy.interpolate.interp2d](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interp2d.html#scipy.interpolate.interp2d)

Yep, sadly we cannot update yet as there have build time issues, I see the irony bringing this up on this issue 😉 (I have a PR open you can look for if you are interested.)

**Comment by tupui** [comment]
BTW @j-bowhay if you want a challenging issue to work on for the doc. This would be extremely helpful! This is preventing both NumPy and us to update.

**Comment by j-bowhay** [comment]
> BTW @j-bowhay if you want a challenging issue to work on for the doc. This would be extremely helpful! This is preventing both NumPy and us to update.

If it is not fixed by the summer I should hopefully have some time to take a look
